examples/ota-provider-app/esp32/main/include/BdxOtaSender.h

```diff
@@ -0,0 +1,103 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <protocols/bdx/BdxTransferSession.h>
+#include <protocols/bdx/TransferFacilitator.h>
+
+#pragma once
+
+enum BdxSenderErrorTypes
+{
+    kErrorBdxSenderNoError = 0,
+    kErrorBdxSenderStatusReceived,
+    kErrorBdxSenderInternal,
+    kErrorBdxSenderTimeOut,
+};
+
+/**
+ * @brief
+ *   This callback is called when bdx session receives a BlockQuery message.
+ *
+ * @param[in]  context  User context
+ * @param[out] blockBuf Preallocated PacketBuffer, implementor should fill the block buf with the data
+ * @param[out] size     Size of data
+ * @param[out] isEof    Flag which tells whether it is final block or not
+ * @param[in]  offset   Offset to read data from
+ *
+ * @return CHIP_NO_ERROR on success; return appropriate error code otherwise
+ */
+typedef CHIP_ERROR (*OnBdxBlockQuery)(void * context, chip::System::PacketBufferHandle & blockBuf, size_t & size, bool & isEof,
+                                      uint32_t offset);
+
+/**
+ * @brief
+ *   This callback is called when bdx session receives BlockAckEOF message
+ *
+ * @param[in] context User contex
+ */
+typedef void (*OnBdxTransferComplete)(void * context);
+
+/**
+ * @brief
+ *   This callback is called when bdx transfer receives StatusReport messages,
+ *   if there is any internal error, or transfer timed out
+ *
+ * @param[in] context User context
+ * @param[in] status Error code
+ */
+typedef void (*OnBdxTransferFailed)(void * context, BdxSenderErrorTypes status);
+
+struct BdxOtaSenderCallbacks
+{
+    chip::Callback::Callback<OnBdxBlockQuery> * onBlockQuery             = nullptr;
```

Comment:
`chip::Callback::Callback` is reference itself so we can directly store the object rather than its pointer. This can also reduce the need for static variables.

Reply:
[`chip::Callback::Callback`](https://github.com/project-chip/connectedhomeip/blob/master/src/lib/core/CHIPCallback.h#L112) do not contain a constructor with no arguments so, it will generate an error `error: use of deleted function`. Probably using a pointer rather than an object here is a good idea. Also, most of this code will be refactored as stated by Carol. I suggest keep this as is for now.

Reply:
[`chip::Callback::Callback`](https://github.com/project-chip/connectedhomeip/blob/master/src/lib/core/CHIPCallback.h#L112) do not contain a constructor with no arguments so, it will generate an error `error: use of deleted function`. Probably using a pointer rather than an object here is a good idea. Also, most of this code will be refactored as stated by Carol. I suggest keep this as is for now.

Reply:
[`chip::Callback::Callback`](https://github.com/project-chip/connectedhomeip/blob/master/src/lib/core/CHIPCallback.h#L112) do not contain a constructor with no arguments so, it will generate an error `error: use of deleted function`. Probably using a pointer rather than an object here is a good idea. Also, most of this code will be refactored as stated by Carol. I suggest keep this as is for now.